### PR TITLE
merge current hdfs improvements to master

### PR DESCRIPTION
**Pull request by rmuir**
We have made a lot of progress on this plugin, I think we should merge back the changes to master and let them run in jenkins.

It has a real hdfs integration test fixture and tests pass on all operating systems, no longer has jar hell, filesystem, configuration, security logic around hadoop are cleaned up, and so on.

Lets get it going in jenkins builds and take it from there...

### CODE REVIEW

**Comment by rmuir** [comment]
related issue: https://github.com/elastic/elasticsearch/issues/15438

File: core/src/main/java/org/elasticsearch/bootstrap/ESPolicy.java [hunk 1]
```diff
@@ -81,10 +83,39 @@ public boolean implies(ProtectionDomain domain, Permission permission) {
             }
         }
 
+        // Special handling for broken Hadoop code: "let me execute or my classes will not load"
+        // yeah right, REMOVE THIS when hadoop is fixed
+        if (permission instanceof FilePermission && "<<ALL FILES>>".equals(permission.getName())) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/bootstrap/ESPolicy.java]: I have no idea if this has a perf impact? How often is this called?

File: core/src/main/java/org/elasticsearch/bootstrap/ESPolicy.java [hunk 2]
```diff
@@ -81,10 +83,39 @@ public boolean implies(ProtectionDomain domain, Permission permission) {
...
+    private static class Rethrower<T extends Throwable> {
+        private void rethrow(Throwable t) throws T {
+            throw (T) t;
+        }
+    }
+
+    /**
+     * Rethrows <code>t</code> (identical object).
+     */
+    private void rethrow(Throwable t) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/bootstrap/ESPolicy.java]: LOL nice trick

File: core/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java [hunk 3]
```diff
@@ -97,6 +96,7 @@ public InputStream readBlob(String name) throws IOException {
     @Override
     public void writeBlob(String blobName, InputStream inputStream, long blobSize) throws IOException {
         final Path file = path.resolve(blobName);
+        // TODO: why is this not specifying CREATE_NEW? Do we really need to be able to truncate existing files?
         try (OutputStream outputStream = Files.newOutputStream(file)) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java]: can't we just put `CREATE_NEW` here? I think there is no need to truncate?

File: plugins/repository-hdfs/build.gradle [hunk 4]
```diff
@@ -18,189 +18,138 @@
...
-        exclude "hadoop-libs/**"
+integTest {
+  boolean fixtureSupported = false;
+  if (Os.isFamily(Os.FAMILY_WINDOWS)) {
+    // hdfs fixture will not start without hadoop native libraries on windows
+    String nativePath = System.getenv("HADOOP_HOME")
+    if (nativePath != null) {
+      Path path = Paths.get(nativePath);
+      if (Files.isDirectory(path) &&
+          Files.exists(path.resolve("bin").resolve("winutils.exe")) &&
```
Comment (s1monw) [inline plugins/repository-hdfs/build.gradle]: oh boy :)

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobStore.java [hunk 5]
```diff
@@ -18,92 +18,84 @@
...
+final class HdfsBlobStore implements BlobStore {
+
+    private final Path root;
+    private final FileContext fileContext;
+    private final int bufferSize;
+    private volatile boolean closed;
+
+    HdfsBlobStore(FileContext fileContext, String path, int bufferSize) throws IOException {
+        this.fileContext = fileContext;
+        this.bufferSize = bufferSize;
```
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobStore.java]: can we somehow assert that buffer-size is not ridiculous?

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsPlugin.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    @SuppressForbidden(reason = "Needs a security hack for hadoop on windows, until HADOOP-XXXX is fixed")
+    private static Void evilHadoopInit() {
+        // hack: on Windows, Shell's clinit has a similar problem that on unix,
+        // but here we can workaround it for now by setting hadoop home
+        // on unix: we still want to set this to something we control, because
+        // if the user happens to have HADOOP_HOME in their environment -> checkHadoopHome goes boom
+        // TODO: remove THIS when hadoop is fixed
+        Path hadoopHome = null;
+        String oldValue = null;
+        try {
```
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsPlugin.java]: man how broken is this that you have to do all this shit to make it even work?

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsPlugin.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            hadoopHome = Files.createTempDirectory("hadoop").toAbsolutePath();
+            oldValue = System.setProperty("hadoop.home.dir", hadoopHome.toString());
+            Class.forName("org.apache.hadoop.security.UserGroupInformation");
+            Class.forName("org.apache.hadoop.util.StringUtils");
+            Class.forName("org.apache.hadoop.util.ShutdownHookManager");
+            Class.forName("org.apache.hadoop.conf.Configuration");
+            Class.forName("org.apache.hadoop.hdfs.protocol.HdfsConstants");
+            Class.forName("org.apache.hadoop.hdfs.protocol.datatransfer.PipelineAck");
+        } catch (ClassNotFoundException | IOException e) {
+            throw new RuntimeException(e);
```
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsPlugin.java]: This warrants a new exception `throw new FUBAR()`?

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java [hunk 8]
```diff
@@ -31,202 +42,114 @@
...
-    // as the FileSystem is long-lived and might go away, make sure to check it before it's being used.
     @Override
-    public FileSystem getFileSystem() throws IOException {
-        SecurityManager sm = System.getSecurityManager();
-        if (sm != null) {
-            // unprivileged code such as scripts do not have SpecialPermission
-            sm.checkPermission(new SpecialPermission());
+    protected void doStart() {
+        String uriSetting = repositorySettings.settings().get("uri");
+        if (!Strings.hasText(uriSetting)) {
```
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: nit: `== false`?

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java [hunk 9]
```diff
@@ -31,202 +42,114 @@
...
-            }
-            if (th instanceof RuntimeException) {
-                throw (RuntimeException) th;
-            }
-            if (th instanceof IOException) {
-                throw (IOException) th;
-            }
-            throw new ElasticsearchException(pae);
+        URI uri = URI.create(uriSetting);
+        if (!"hdfs".equalsIgnoreCase(uri.getScheme())) {
```
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: nit: `== false`

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java [hunk 10]
```diff
@@ -31,202 +42,114 @@
...
-            ClassLoader oldCL = th.getContextClassLoader();
-            try {
-                th.setContextClassLoader(getClass().getClassLoader());
-                return initFileSystem(repositorySettings);
-            } catch (IOException ex) {
-                throw ex;
-            } finally {
-                th.setContextClassLoader(oldCL);
+        
+        int bufferSize = (int) repositorySettings.settings().getAsBytesSize("buffer_size", new ByteSizeValue(100, ByteSizeUnit.KB)).bytes();
```
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: you can call `bytesAsInt()` then you safe the cast and it checks that it's lossless

File: core/src/main/java/org/elasticsearch/bootstrap/ESPolicy.java (hunk 1 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/ESPolicy.java]: this permission is only requested by hadoop, it is the only thing really trying to execute like that.

File: core/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java (hunk 3 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java]: Tests rely on this broken behavior, so i opened https://github.com/elastic/elasticsearch/issues/15579 for it. 

HDFS plugin acts correctly and passes the equivalent of CREATE_NEW

File: test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java [hunk 11]
```diff
@@ -119,25 +138,42 @@ protected boolean resetNodeAfterTest() {
...
+
+    /** Helper method to create list of plugins without specifying generic types. */
+    @SafeVarargs
+    @SuppressWarnings("varargs") // due to type erasure, the varargs type is non-reifiable, which casues this warning
+    protected final Collection<Class<? extends Plugin>> pluginList(Class<? extends Plugin>... plugins) {
+        return Arrays.asList(plugins);
+    }
+
+    private Node newNode() {
+        Settings settings = Settings.builder()
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java]: looks good - thanks for the cleanup

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java (hunk 10 above)
Comment (rmuir) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: OK, I actually have no clue what this setting is, or what it is doing: its not new in this PR, its just the line of code is moved from somewhere else...

**Comment by s1monw** [comment]
I left some really minor comments as I went through the code. Having the fixture for HDFS is a huge improvement and getting the stuff actually working is awesome. I am surprised you got that far with all these hacks? it's pretty crazy what's happening in HDFS!!!

LGTM - please take one more look at the minor comments and push once you are done

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java (hunk 10 above)
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: it getting "buffer_size" and if it's not there it uses the 100 KB as a default, we should make the default a constant?

Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: 100kb also seems arbitrary, I know it was here before maybe @costin knows

**Comment by rmuir** [comment]
ok lets destroy jenkins!

**Pull request merged**